components/ws-daemon/pkg/daemon/cgroup_customizer.go

```diff
@@ -64,8 +78,72 @@ func (c *CgroupCustomizer) WorkspaceAdded(ctx context.Context, ws *dispatch.Work
 	return nil
 }
 
-func (c *CgroupCustomizer) customV1() ([]cgroups.Subsystem, error) {
+func (c *CgroupCustomizerV1) customV1() ([]cgroups.Subsystem, error) {
 	return []cgroups.Subsystem{
 		cgroups.NewDevices(c.cgroupBasePath),
 	}, nil
 }
+
+type CgroupCustomizerV2 struct {
+	cgroupBasePath string
+}
+
+func (c *CgroupCustomizerV2) WorkspaceAdded(ctx context.Context, ws *dispatch.Workspace) error {
+	cgroupPath, err := retrieveCGroupPath(ctx, ws.ContainerID)
+	if err != nil {
+		return err
+	}
+
+	fullCgroupPath := filepath.Join(c.cgroupBasePath, cgroupPath)
+	cgroupFD, err := unix.Open(fullCgroupPath, unix.O_DIRECTORY|unix.O_RDONLY, 0o600)
+	if err != nil {
+		return xerrors.Errorf("cannot get directory fd for %s", fullCgroupPath)
+	}
+	defer unix.Close(cgroupFD)
+
+	insts, license, err := devicefilter.DeviceFilter(composeDeviceRules())
+	if err != nil {
+		return xerrors.Errorf("failed to generate device filter: %w", err)
+	}
+
+	_, err = ebpf.LoadAttachCgroupDeviceFilter(insts, license, cgroupFD)
```

Comment:
If there is an err, it is necessary to call `closer` of the return value. Or you may need to revert back to the original ebpf program. the `closer` will only detach and will not revert the original ebpf code. Also, I should be a bit concerned that the workspace will probably stop working if this code fails, but I don't have a good idea... 😭 
@Furisto @csweichel Are there any good ideas?

Reply:
What about applying it to the pod cgroup instead of the container cgroup and erasing the ebpf in the container cgroup? Maybe elimination may not be necessary due to inheritance from parents. In this case, even if it fails, the ebpf code for the container's cgroup device will remain, so there should be no such thing as the user being able to access all devices.

Reply:
There are two cases with the closer:
- There is only one ebpf program attached to the cgroup (this is the current case). In this case the closer is a dummy and does not do anything -> No need to call the closer
- There are multiple ebpf programs attached to the cgroup. In this case the closer will detach the program we have attached when an error occurs while detaching the other programs. That is the order is -> attach new program -> setup closer -> detach all other programs. In that case I would not want to call the closer either because it could mean that we have detached the old runc device rules, then some error happens while detaching another program and then if we call the closer we would detach our newly attached program leaving us with no device rules to protect us.

> Also, I should be a bit concerned that the workspace will probably stop working if this code fails, but I don't have a good > idea... sob

Why would the workspace stop working?

My understanding was that ebpf programs are atomically replaced, so either the old program is attached or the new program,

Reply:
> My understanding was that ebpf programs are atomically replaced, so either the old program is attached or the new program.

I have asked in the ebpf slack and got the confirmation that this is indeed the behavior.

Reply:
Oops, I was reading the code backward for detach and attach in the old eBPF code. My brain seems to be not good today. Looks like it's not a problem.
https://github.com/opencontainers/runc/blob/067aaf8548d78269dcb2c13b856775e27c410f9c/libcontainer/cgroups/ebpf/ebpf_linux.go#L242